runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -7294,17 +7298,120 @@ done:;
 	/*
 	 * Determine if the two valueTypes are substitutable when rhs.class equals lhs.class
 	 *
-	 * @param[in] lhs the lhs object of acmp bytecodes and it's a valueType
-	 * @param[in] rhs the rhs object of acmp bytecodes and it's a valueType
+	 * @param[in] lhs the lhs object address
+	 * @param[in] rhs the rhs object address
+	 * @param[in] startOffset the initial offset for the object
+	 * @param[in] clazz the value type class
 	 * return true if they are substitutable and false otherwise
 	 */
-	VMINLINE bool
-	isSubstitutable(j9object_t lhs, j9object_t rhs)
-	{
-		/*
-		 * TODO: this will be updated in a future PR.
-		 */
-		return false;
+	bool
+	isSubstitutable(j9object_t lhs, j9object_t rhs, UDATA startOffset, J9Class *clazz)
+	{
+		U_32 walkFlags = J9VM_FIELD_OFFSET_WALK_INCLUDE_INSTANCE;
+		J9ROMFieldOffsetWalkState state;
+		J9ROMFieldOffsetWalkResult *result = fieldOffsetsStartDo(_vm, clazz->romClass, VM_VMHelpers::getSuperclass(clazz), &state, walkFlags, clazz->flattenedClassCache);
+
+		while (NULL != result->field) {
+			J9UTF8 *signature = J9ROMNAMEANDSIGNATURE_SIGNATURE(&result->field->nameAndSignature);
+			U_8 *sigChar = J9UTF8_DATA(signature);
+
+			switch (*sigChar) {
+			case 'Z': /* Boolean */
+			case 'B': /* Byte */
+			case 'C': /* Char */
+			case 'I': /* Int */
+			case 'S': { /* Short */ 
+				I_32 lhsValue = _objectAccessBarrier.inlineMixedObjectReadI32(_currentThread, lhs, startOffset + result->offset);
+				I_32 rhsValue = _objectAccessBarrier.inlineMixedObjectReadI32(_currentThread, rhs, startOffset + result->offset);
+				if (lhsValue != rhsValue) {
+					return false;
+				}
+				break;
+			}
+			case 'J': { /* Long */
+				I_64 lhsValue = _objectAccessBarrier.inlineMixedObjectReadI64(_currentThread, lhs, startOffset + result->offset);
+				I_64 rhsValue = _objectAccessBarrier.inlineMixedObjectReadI64(_currentThread, rhs, startOffset + result->offset);
+				if (lhsValue != rhsValue) {
+					return false;
+				}
+				break;
+			}
+			case 'D': { /* Double */
+				U_64 lhsValue = _objectAccessBarrier.inlineMixedObjectReadU64(_currentThread, lhs, startOffset + result->offset);
+				U_64 rhsValue = _objectAccessBarrier.inlineMixedObjectReadU64(_currentThread, rhs, startOffset + result->offset);
+				I_32 result = helperDoubleCompareDouble((jdouble *)&lhsValue, (jdouble *)&rhsValue);
```

Comment:
there is an issue with the floating point comparisons, it should behave like `https://docs.oracle.com/javase/7/docs/api/java/lang/Float.html#equals(java.lang.Object)`

Reply:
same for float